test/fpm_test/test_source_parsing.f90

```diff
@@ -380,6 +382,96 @@ subroutine test_program(error)
 
     end subroutine test_program
 
+    !> Try to parse a simple fortran program with no "program" header
+    subroutine test_program_noheader(error)
+
+        !> Error handling
+        type(error_t), allocatable, intent(out) :: error
+
+        integer :: unit
+        character(:), allocatable :: temp_file
+        type(srcfile_t), allocatable :: f_source
+
+        allocate(temp_file, source=get_temp_filename())
+
+        open(file=temp_file, newunit=unit)
+        write(unit, '(a)') &
+            & 'use program_one', &
+            & 'implicit none', &
+            & 'integer :: module, program', &
+            & 'module = 1', &
+            & 'module= 1', &
+            & 'module =1', &
+            & 'module (i) =1', &
+            & 'program = 123', &
+            & 'contains', &
+            & 'subroutine f()', &
+            & 'end subroutine f', &
+            & 'end program my_program'
```

Comment:
This would be an invalid Fortran program. The program name can only appear in an end program statement, if there is a program statement at the top.